manuals/olecfexport.1
.Dd April 25, 2019
.Dt olecfexport
.Os libolecf
.Sh NAME
.Nm olecfexport
.Nd exports items stored in a Object Linking and Embedding (OLE) Compound File (CF)
.Sh SYNOPSIS
.Nm olecfexport
.Op Fl c Ar codepage
.Op Fl f Ar format
.Op Fl l Ar logfile
.Op Fl m Ar mode
.Op Fl t Ar target
.Op Fl dhqvV
.Ar source
.Sh DESCRIPTION
.Nm olecfexport
is a utility to export items stored in a Object Linking and Embedding (OLE) Compound File (CF)
.Pp
.Nm olecfexport
is part of the
.Nm libolecf
package.
.Nm libolecf
is a library to access the Object Linking and Embedding (OLE) Compound File (CF) format
.Pp
.Ar source
is the source file.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl c Ar codepage
specify the codepage of ASCII strings, options: ascii, windows-874, windows-932, windows-936, windows-949, windows-950, windows-1250, windows-1251, windows-1252 (default), windows-1253, windows-1254, windows-1255, windows-1256, windows-1257 or windows-1258
.It Fl d
dumps the item values in a separate file: ItemValues.txt
.It Fl f Ar format
specify the preferred output format, options: all, html, rtf, text (default)
.It Fl h
shows this help
.It Fl l Ar logfile
specify the file in which to log information about the exported items
.It Fl m Ar mode
export mode, option: all, debug, items (default), recovered. 'all' exports the (allocated) items, orphan and recovered items. 'debug' exports all the (allocated) items, also those outside the the root folder. 'items' exports the (allocated) items. 'recovered' exports the orphan and recovered items.
.It Fl q
quiet shows minimal status information
.It Fl t Ar target
specify the basename of the target directory to export to (default is the source filename followed by .export)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# olecfexport file.doc
olecfexport 20101027
.sp
.Dl ...
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh BUGS
Please report bugs of any kind to <joachim.metz@gmail.com> or on the project website:
https://github.com/libyal/libolecf/
.Sh AUTHOR
These man pages were written by Joachim Metz.
.Sh COPYRIGHT
Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>.
This is free software; see the source for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.Sh SEE ALSO
.Xr olecfinfo 1


ChangeLog
TODO
* add support for variant type and multi variant type
* double check format bounds checks
* add API support for property types
  - property values allow to set ascii codepage
  - add multi values support e.g. 0x101e
* use libfvalue table for property values in a property set section?
* fix: info_handle_document_summary_information_fprint: unable to read property set stream
* libolecf:
  - restrict stream functions to stream directory types
  - add stream read random
  - libolecf_stream_read_buffer change uint8_t* to void* ?
  - fill unallocated_block_list
* pyolecf:
  - add format version function
  - property values allow to set ascii codepage
  - pyolecf_property_value add functions
  - fix pyolecf_property_value get data function
  - add more definitions types
* olecfexport
  - add stream name to export failed error, if possible
* olecfmount
  - sanitize filenames
* support files with an empty root directory entry
* update man pages
  - correct olecfinfo.1
  - correct olecfexport.1
  - add olecfmount.1
* add support for multi types (word2003 test file)
* refactor libolecf_io_handle_read_stream into libolecf_io_handle_read_stream ?
* use libfdata vector to cache sectors ?
* olecftools: check UTF16 print support
* implement
  - libolecf_file_get_number_of_unallocated_blocks
  - libolecf_file_get_unallocated_block

* implement stream IO
* implement (document) summary information stream support
  implement property set stream support

* implement Workbook stream support
  in separate library
* implement olecf msg file support (Outlook MSG file format)
  in separate library

TEST
* add automated tests
* files with multi sector MSAT

API
* file io handle functions
* file information functions
* stream support
* allocation data support
* compound object support
* property set support

Library
* clone file io handle in item
* directory entry
  - debug print node color description
* property set stream
  - implement api functions and internal structures
  - debug print remnant and trailing data
  - debug print property names and values
* object stream
  - implement api functions and internal structures
* code clean up
  - check for code where list functions should be used instead
  - check for code where tree functions should be used instead

TOOLS
olecfinfo
* print basic file data (partial)
* print allocation data
* print stream/directory data (partial)
* print compobj stream info

20140930
* see `git log' for more recent change log
* removed README.macosx
* changes for project site move

20140810
* worked on libfwps support

20140801
* bug fix in Python-bindings

20140727
* worked on olecfmount Dokan support

20140726
* updated dependencies

20140329
* small changes to Makefile

20140323
* worked on Python bindings

20140317
* worked on setup.py

20140301
* 2014 update
* updated dependencies
* worked on Python bindings

20131209
* updated dependencies

20131108
* added support for 0 value sub directory entry not being used as index.

20131012
* added format version functions

20131011
* changes for non 512 sector sized files

20131009
* changes to dpkg files

20131008
* improved pyolecf as integer functions

20130929
* worked on setup.py, largely for MSI builds

20120928
* updated libfvalue and libfdatetime for filetime to integer functionality

20120925
* worked on API

20120924
* worked on pyolecf
* updated msvscpp files

20120923
* code clean up
* fixed memory leak in olecfexport for 0 byte streams

20120918
* worked on API
* olecfinfo: worked on property values

20120917
* code clean up
* worked on tests
* olecfmount: added extended options
* worked on API
* olecfinfo: worked on property values

20120916
* updated dependencies

20120717
* updated dependencies
* worked on tests
* olecfexport: fixed .export suffix

20120609
* updated dependencies

20120421
* worked on tests

20120414
* Textual changes
* worked on tests
* updates and bug fixes in pyolecf
* fixed codepage 949, 950 and 1255 restriction

20120413
* updated dependencies

20120319
* updated dependencies

20130305
* worked on libcdata integration
* worked on libfole

20130303
* worked on PackageMaker files

20130219
* worked on python bindings

20130218
* worked on python bindings

20130217
* fixed bug introduced in libcdata update
* worked on python bindings
* code clean up
* updated msvscpp files
* fixes in pc and spec file

20130211
* small changes

20130210
* 2013 update
* updated olecfmount added -X
* fixed olecfmount reading files
* worked on libcdata changes
* updated dependencies

20121104
* updated dependencies

20120920
* worked on libcdata changes

20120919
* updated dependencies

20120607
* olecfexport
 - worked on moving olecf file into export handle
* worked on using libfvalue for property values

20120530
* removed libcsystem_file_stream from export handle
  replaced by libcfile file
* code clean up remove item attach and detach functions

20120526
* fixes for olecfinfo requiring SummaryInformation and
  DocumentSummaryInformation directory entries
* updated dependencies
* worked on olecfmount
* added functions to retrieve item creation and modification time
* added missing libolecf_item_get_sub_item_by_utf16_name

20120523
* updated dependencies
* changes due to project move

20120508
* updated dependencies
* code clean up

20120503
* fixed issue in olecfexport regarding make directory
* applied updates

20120102
* 2012 update

20111023
* updated configure.ac and m4 scripts
* updated README files
* updated spec an pc file
* updated common, array type, libsystem, libuna, libbfio, libfdatetime,
  libfvalue

20110704
* worked on multi value support
* added libfvalue

20110703
* changes for: warning: AC_LANG_CONFTEST: no AC_LANG_SOURCE call detected in body
* updated libuna, libbfio, libfdatetime
* refactored property set reading code
* added debug string for directory entry type
* added codepage functionality for property set streams

20110623
* code clean up
* updated olecfiniput.[ch]

olecftools/olecfexport.c
/*
 * Exports streams from an OLE Compound File
 *
 * Copyright (C) 2008-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <memory.h>
#include <narrow_string.h>
#include <system_string.h>
#include <types.h>
#include <wide_string.h>

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#if defined( HAVE_STDLIB_H )
#include <stdlib.h>
#endif

#include <stdio.h>

#include "export_handle.h"
#include "log_handle.h"
#include "olecftools_getopt.h"
#include "olecftools_libcerror.h"
#include "olecftools_libclocale.h"
#include "olecftools_libcnotify.h"
#include "olecftools_libcpath.h"
#include "olecftools_libolecf.h"
#include "olecftools_output.h"
#include "olecftools_signal.h"
#include "olecftools_unused.h"

export_handle_t *olecfexport_export_handle = NULL;
int olecfexport_abort                      = 0;

/* Prints the executable usage information
 */
void usage_fprint(
      FILE *stream )
{
	if( stream == NULL )
	{
		return;
	}
	fprintf( stream, "Use olecfexport to export streams from an OLE Compound File.\n\n" );

	fprintf( stream, "Usage: olecfexport [ -c codepage ] [ -l logfile ] [ -t target ]\n"
	                 "                   [ -hvV ] source\n\n" );

	fprintf( stream, "\tsource: the source file\n\n" );

	fprintf( stream, "\t-c:     codepage of ASCII strings, options: ascii, windows-874,\n"
	                 "\t        windows-932, windows-936, windows-949, windows-950,\n"
	                 "\t        windows-1250, windows-1251, windows-1252 (default),\n"
	                 "\t        windows-1253, windows-1254, windows-1255, windows-1256,\n"
	                 "\t        windows-1257 or windows-1258\n" );
	fprintf( stream, "\t-h:     shows this help\n" );
	fprintf( stream, "\t-l:     logs information about the exported items\n" );
	fprintf( stream, "\t-t:     specify the target directory to export to\n"
	                 "\t        (default is the source filename followed by .export)\n" );
	fprintf( stream, "\t-v:     verbose output to stderr\n" );
	fprintf( stream, "\t-V:     print version\n" );
}

/* Signal handler for olecfexport
 */
void olecfexport_signal_handler(
      olecftools_signal_t signal OLECFTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "olecfexport_signal_handler";

	OLECFTOOLS_UNREFERENCED_PARAMETER( signal )

	olecfexport_abort = 1;

	if( olecfexport_export_handle != NULL )
	{
		if( export_handle_signal_abort(
		     olecfexport_export_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal export handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	libcerror_error_t *error                  = NULL;
	log_handle_t *log_handle                  = NULL;
	system_character_t *log_filename          = NULL;
	system_character_t *option_ascii_codepage = NULL;
	system_character_t *option_target_path    = NULL;
	system_character_t *path_separator        = NULL;
	system_character_t *source                = NULL;
	char *program                             = "olecfexport";
	system_integer_t option                   = 0;
	size_t source_length                      = 0;
	int result                                = 0;
	int verbose                               = 0;

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "olecftools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( olecftools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	olecftools_output_version_fprint(
	 stdout,
	 program );

	while( ( option = olecftools_getopt(
	                   argc,
	                   argv,
	                   _SYSTEM_STRING( "c:hl:t:vV" ) ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				usage_fprint(
				 stdout );

				return( EXIT_FAILURE );

			case (system_integer_t) 'c':
				option_ascii_codepage = optarg;

				break;

			case (system_integer_t) 'h':
				usage_fprint(
				 stdout );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'l':
				log_filename = optarg;

				break;

			case (system_integer_t) 't':
				option_target_path = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				olecftools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source file.\n" );

		usage_fprint(
		 stdout );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	if( option_target_path == NULL )
	{
		source_length = system_string_length(
		                 source );

		path_separator = system_string_search_character_reverse(
		                  source,
		                  (system_character_t) LIBCPATH_SEPARATOR,
		                  source_length );

		if( path_separator == NULL )
		{
			path_separator = source;
		}
		else
		{
			path_separator++;
		}
		option_target_path = path_separator;
	}
	libcnotify_verbose_set(
	 verbose );
	libolecf_notify_set_stream(
	 stderr,
	 NULL );
	libolecf_notify_set_verbose(
	 verbose );

	if( export_handle_initialize(
	     &olecfexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize export handle.\n" );

		goto on_error;
	}
	if( export_handle_set_target_path(
	     olecfexport_export_handle,
	     option_target_path,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to set target path.\n" );

		goto on_error;
	}
	result = export_handle_create_items_export_path(
	          olecfexport_export_handle,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to create items export path.\n" );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "%" PRIs_SYSTEM " already exists.\n",
		 olecfexport_export_handle->items_export_path );

		goto on_error;
	}
	if( log_handle_initialize(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize log handle.\n" );

		goto on_error;
	}
	if( log_handle_open(
	     log_handle,
	     log_filename,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	if( option_ascii_codepage != NULL )
	{
		result = export_handle_set_ascii_codepage(
		          olecfexport_export_handle,
		          option_ascii_codepage,
		          &error );

		if( result == -1 )
		{
			fprintf(
			 stderr,
			 "Unable to set ASCII codepage in export handle.\n" );

			goto on_error;
		}
		else if( result == 0 )
		{
			fprintf(
			 stderr,
			 "Unsupported ASCII codepage defaulting to: windows-1252.\n" );
		}
	}
	fprintf(
	 stdout,
	 "Opening file.\n" );

	if( export_handle_open_input(
	     olecfexport_export_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	if( export_handle_export_file(
	     olecfexport_export_handle,
	     log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to export file stream and storage items.\n" );

		goto on_error;
	}
	if( export_handle_close(
	     olecfexport_export_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close export handle.\n" );

		goto on_error;
	}
	if( export_handle_free(
	     &olecfexport_export_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free export handle.\n" );

		goto on_error;
	}
	if( log_handle_close(
	     log_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close log file: %" PRIs_SYSTEM ".\n",
		 log_filename );

		goto on_error;
	}
	if( log_handle_free(
	     &log_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free log handle.\n" );

		goto on_error;
	}
	fprintf(
	 stdout,
	 "Export completed.\n" );

	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( log_handle != NULL )
	{
		log_handle_close(
		 log_handle,
		 NULL );
		log_handle_free(
		 &log_handle,
		 NULL );
	}
	if( olecfexport_export_handle != NULL )
	{
		export_handle_close(
		 olecfexport_export_handle,
		 NULL );
		export_handle_free(
		 &olecfexport_export_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

